tensorbay/cli/log.py

```diff
@@ -24,19 +25,111 @@
 """
 
 
-def _implement_log(
+class _CommitNode:  # pylint: disable=too-few-public-methods
+    def __init__(self, commit: Commit):
+        self.commit: Commit = commit
+        self.child_commits: List[_CommitNode] = list()
+        self.available_child_num = 0
+        self.parent_commit: _CommitNode
+        self.branch_names: Set[Optional[str]] = set()
+
+
+def _build_commit_tree(
+    dataset_client: DatasetClientType, valid_branch_name: List[Optional[str]]
+) -> Tuple[Optional[_CommitNode], List[_CommitNode]]:
+    root = None
+    commits_dict: Dict[Optional[str], _CommitNode] = dict()
+    leaf_set: Set[_CommitNode] = set()
+    for branch_name in valid_branch_name:
+        commits = dataset_client.list_commits(branch_name)
+        last_node, leaf_flag = None, True
+        for commit in commits:
+            # Commit already exists in the tree.
+            if commit.commit_id in commits_dict.keys():
+                commits_dict[commit.commit_id].branch_names.add(branch_name)
+                break
+            # Commit does not exists in the tree.
+            current_node = _CommitNode(commit)
+            current_node.branch_names.add(branch_name)
+            # Save leaf node to leaf set.
+            if leaf_flag:
+                leaf_set.add(current_node)
+                leaf_flag = False
+            # Commit has parent commit node.
+            if last_node:
+                last_node.parent_commit = current_node  # type: ignore[unreachable]
+                current_node.child_commits.append(last_node)
+                current_node.available_child_num += 1
+            # Save commit to commit dict.
+            commits_dict[commit.commit_id] = current_node
+            # root's parent commit id is "000..."
+            if commit.parent_commit_id == ROOT_COMMIT_ID:
+                root = current_node
+                break
+            parent_commit_id = commit.parent_commit_id
+            # Parent node has existed in the tree.
+            if parent_commit_id in commits_dict.keys():
+                parent_node: _CommitNode = commits_dict[parent_commit_id]
+                parent_node.child_commits.append(current_node)
+                parent_node.available_child_num += 1
+                parent_node.branch_names.add(branch_name)
+                current_node.parent_commit = commits_dict[parent_commit_id]
```

Comment:
Could this be covered by L48-50 and L59-62?

Reply:
No, I think not.~